iocore/net/quic/QUICPinger.cc

```diff
@@ -27,62 +27,59 @@ void
 QUICPinger::request(QUICEncryptionLevel level)
 {
   SCOPED_MUTEX_LOCK(lock, this->_mutex, this_ethread());
-  if (!this->_is_level_matched(level)) {
-    return;
-  }
   ++this->_need_to_fire[static_cast<int>(level)];
 }
 
 void
 QUICPinger::cancel(QUICEncryptionLevel level)
 {
   SCOPED_MUTEX_LOCK(lock, this->_mutex, this_ethread());
-  if (!this->_is_level_matched(level)) {
-    return;
-  }
-
   if (this->_need_to_fire[static_cast<int>(level)] > 0) {
     --this->_need_to_fire[static_cast<int>(level)];
   }
 }
 
 bool
-QUICPinger::will_generate_frame(QUICEncryptionLevel level, uint32_t seq_num)
+QUICPinger::_will_generate_frame(QUICEncryptionLevel level, size_t current_packet_size, bool ack_eliciting)
 {
   SCOPED_MUTEX_LOCK(lock, this->_mutex, this_ethread());
-  if (!this->_is_level_matched(level)) {
+  // PING Frame is meaningless for ack_eliciting packet. Cancel it.
+  if (ack_eliciting) {
+    this->_ack_eliciting_packet_out = true;
+    this->cancel(level);
     return false;
   }
 
-  if (seq_num == this->_latest_seq_num) {
-    return false;
+  if (this->_ack_eliciting_packet_out == false && !ack_eliciting && current_packet_size > 0 &&
+      this->_need_to_fire[static_cast<int>(level)] == 0) {
+    // force to send an PING Frame
+    this->request(level);
   }
 
-  this->_latest_seq_num = seq_num;
+  // only update `_ack_eliciting_packet_out` when we has something to send.
+  if (current_packet_size) {
```

Comment:
Add this check to avoid un-nessesary ping triggered.

Reply:
This is fine at all. I think it's a right direction.